ledger/ledger-common/src/main/kotlin/net/corda/v5/ledger/common/Amount.kt

```diff
@@ -0,0 +1,366 @@
+ * @property quantity The quantity of the item of the current amount.
+ * @property item The item being quantified by the current amount.
+ * @property isDivisible Determines whether the current amount can be subdivided into smaller amounts.
+ */
+@CordaSerializable
+data class Amount<T : Any>(val quantity: BigInteger, val item: T) : Comparable<Amount<T>> {
+
+    /**
+     * Creates an instance of the [Amount] class.
+     *
+     * @param quantity The quantity of the item of the current amount.
+     * @param item The item being quantified by the current amount.
+     */
+    constructor(quantity: Int, item: T) : this(quantity.toBigInteger(), item)
+
+    /**
+     * Creates an instance of the [Amount] class.
+     *
+     * @param quantity The quantity of the item of the current amount.
+     * @param item The item being quantified by the current amount.
+     */
+    constructor(quantity: Long, item: T) : this(quantity.toBigInteger(), item)
+
+    /**
+     * Creates an instance of the [Amount] class.
+     *
+     * Note that [BigDecimal] is converted to an unscaled [BigInteger]
+     *
+     * @param quantity The quantity of the item of the current amount.
+     * @param item The item being quantified by the current amount.
+     */
+    constructor(quantity: BigDecimal, item: T) : this(quantity.unscaledValue(), item)
+
+    companion object {
+
+        /**
+         * Creates an [Amount] of zero of the specified item.
+         *
+         * @param T The underlying type of the item.
+         * @property item The item being quantified.
+         * @return Returns an [Amount] of zero of the item.
+         */
+        @JvmStatic
+        fun <T : Any> zero(item: T): Amount<T> {
+            return Amount(BigInteger.ZERO, item)
+        }
+
+        /**
+         * Creates an [Amount] of one of the specified item.
+         *
+         * @param T The underlying type of the item.
+         * @property item The item being quantified.
+         * @return Returns an [Amount] of one of the item.
+         */
+        @JvmStatic
+        fun <T : Any> one(item: T): Amount<T> {
+            return Amount(BigInteger.ONE, item)
+        }
+
+        @JvmStatic
+        fun fromCurrency(quantity: Int, currency: Currency): Amount<Currency> {
+            return fromCurrency(quantity.toBigInteger(), currency)
+        }
+
+        @JvmStatic
+        fun fromCurrency(quantity: Long, currency: Currency): Amount<Currency> {
+            return fromCurrency(quantity.toBigInteger(), currency)
+        }
+
+        @JvmStatic
+        fun fromCurrency(quantity: BigInteger, currency: Currency): Amount<Currency> {
+            return fromCurrency(quantity.toBigDecimal(currency.defaultFractionDigits), currency)
+        }
+
+        @JvmStatic
+        fun fromCurrency(quantity: BigDecimal, currency: Currency): Amount<Currency> {
+            val quantityScale = quantity.scale()
+            val currencyScale = currency.defaultFractionDigits
```

Comment:
There are so many problems with Java Currency that I would be keen to drop this. There are numerous historical cases in which currencies were revalued, or renamed and Java just doesn't have accurate master reference data. That data would have to live on the ledger.

Reply:
Okay happy to drop.